fiat-roles/src/main/groovy/com/netflix/spinnaker/fiat/permissions/RedisPermissionsRepository.java

```diff
@@ -16,7 +16,228 @@
+import java.util.stream.Collectors;
+
+/**
+ * This Redis-backed permission repository is structured in a way to optimized reading types of
+ * resource permissions. In general, this looks like a key schema like:
+ *
+ * "prefix:myuser@domain.org:resources" -> {"resourceName1" -> "<serialized json of resourceName1>",
+ *                                          "resourceName2" -> "<serialized json of resourceName1>"}
+ *
+ * Additionally, a helper key, called the "all users" key, maintains a set of all usernames.
+ */
+@Component
+@Slf4j
+public class RedisPermissionsRepository implements PermissionsRepository {
+
+  private static final String KEY_PERMISSIONS = "permissions";
+  private static final String KEY_ALL_USERS = "users";
+
+  @Value("${fiat.redis.prefix:spinnaker:fiat}")
+  @Setter
+  private String prefix;
+
+  @Autowired
+  @Setter
+  private ObjectMapper objectMapper;
+
+  @Autowired
+  @Setter
+  private JedisSource jedisSource;
+
+  // TODO(ttomsu): Add RedisCacheOptions from Clouddriver.
+
+  @Override
+  public PermissionsRepository put(@NonNull UserPermission permission) {
+    try (Jedis jedis = jedisSource.getJedis()) {
+      for (Resource r : Resource.values()) {
+        Map<String, String> resourceValues = new HashMap<>();
+
+        Consumer<Named> putInValuesMap = namedResource -> {
+          try {
+            resourceValues.put(namedResource.getName(),
+                               objectMapper.writeValueAsString(namedResource));
+          } catch (JsonProcessingException jpe) {
+            log.error("Serialization exception writing " + permission.getId() + " entry.", jpe);
+          }
+        };
+
+        switch (r) {
+          case ACCOUNT:
+            permission.getAccounts().stream().forEach(putInValuesMap);
+            break;
+          case APPLICATION:
+            permission.getApplications().stream().forEach(putInValuesMap);
+            break;
+        }
+
+        if (!resourceValues.isEmpty()) {
+          jedis.hmset(userKey(permission.getId(), r), resourceValues);
+          jedis.sadd(allUsersKey(), permission.getId());
+        }
+      }
+    } catch (Exception e) {
+      log.error("Storage exception writing " + permission.getId() + " entry.", e);
+    }
+    return this;
+  }
+
+  @Override
+  public UserPermission get(@NonNull String id) {
+    UserPermission userPermission = new UserPermission().setId(id);
+    try (Jedis jedis = jedisSource.getJedis()) {
+      for (Resource r : Resource.values()) {
+        Map<String, String> resourceMap = jedis.hgetAll(userKey(id, r));
+
+        switch (r) {
+          case ACCOUNT:
+            userPermission.getAccounts().addAll(extractAccounts(resourceMap));
+            break;
+          case APPLICATION:
+            userPermission.getApplications().addAll(extractApplications(resourceMap));
+            break;
+        }
+      }
+    } catch (Exception e) {
+      log.error("Storage exception reading " + id + " entry.", e);
```

Comment:
I'd probably go with Optional<UserPermission> and forgo the `Optional.ofNullable()` in the callers.